### PR TITLE
New linter rules

### PR DESCRIPTION
**Pull request by luanpotter**
I've been studying both `dart_code_metrics` and our regular linter (that has many new rules) to see what I can learn.I basically added all rules that made sense and weren't too hard to fix. Some rules have names that make sense but the implementation is too broad to be useful. The ones I pick I think make a lot of sense. Some I want to add but I will do after the examples are removed (for example, forbid print statements).

The major changes are:

* imports ordering the way I specify on styleguide (but sometimes fail to check on PRs) is now a rule \o/
* local variables are never explicitly typed. fields/methods are always explicitly typed. (we had almost that but there were a few cases left).
* Change all constants to camel case as that is the new dart standard.
* Comments referencing: using [foo] on comments actually creates a doc link. But not everything works. I had to remove some that were broken because the referenced class was not in context. However this rule let me fix A LOT of linking errors that were just typos or outdated references. Also I took the liberty to review all comments I passed through and fix unrelated typos and some outdated stuff. In fact since vscode also highlights grammar errors I fixed a lot of typos everywhere I found.
* Ah finally enforced single quotes that I always have to comment on PRs as well.

Note: this **is** a breaking change because of the constants rename.

### CODE REVIEW

**Review by wolfenrain — APPROVED** [review]
I approve personally of the linter rules, did a quick scan through the changed code. Not in-depth as I assume you just followed the suggesties recommended by said rules :D

File: analysis_options.yaml [hunk 1]
```diff
@@ -15,69 +15,130 @@ linter:
     - always_put_control_body_on_new_line
     - always_require_non_null_named_parameters
     - annotate_overrides
+    - avoid_double_and_int_checks
+    - avoid_dynamic_calls
     - avoid_empty_else
+    - avoid_equals_and_hash_code_on_mutable_classes
```
Comment (spydon) [inline analysis_options.yaml:21]: Why does one want to avoid this?

File: analysis_options.yaml [hunk 2]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - camel_case_types
     - cancel_subscriptions
+    - cast_nullable_to_non_nullable
+    - close_sinks
+    - comment_references
+    - constant_identifier_names
     - control_flow_in_finally
+    - curly_braces_in_flow_control_structures
+    - directives_ordering
+    - do_not_use_environment
```
Comment (spydon) [inline analysis_options.yaml:47]: What does this mean? Not using env variables?

File: analysis_options.yaml [hunk 3]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - iterable_contains_unrelated_type
+    - join_return_with_assignment
     - library_names
     - library_prefixes
     - list_remove_unrelated_type
+    - literal_only_boolean_expressions
+    - missing_whitespace_between_adjacent_strings
     - no_adjacent_strings_in_list
     - no_duplicate_case_values
-    - non_constant_identifier_names
```
Comment (spydon) [inline analysis_options.yaml:40]: Why was this removed?

File: analysis_options.yaml [hunk 4]
```diff
@@ -15,69 +15,130 @@ linter:
...
+    - unnecessary_statements
+    - unnecessary_string_escapes
+    - unnecessary_string_interpolations
     - unnecessary_this
-    - unrelated_type_equality_checks
+    - use_full_hex_values_for_flutter_colors
+    - use_function_type_syntax_for_parameters
+    - use_is_even_rather_than_modulo
     - use_rethrow_when_possible
-    - unnecessary_new
```
Comment (spydon) [inline analysis_options.yaml:76]: Why was this removed?

File: doc/examples/sprites/lib/main_base64.dart [hunk 5]
```diff
@@ -25,7 +24,7 @@ class MyGame extends Game {
   }
 
   @override
-  void update(dt) {}
+  void update(double dt) {}
```
Comment (spydon) [inline doc/examples/sprites/lib/main_base64.dart:27]: I was thinking of this dt vs. t, since we are using both. I vote for that we have them all as dt.

**Review by spydon — COMMENTED** [review]
Looking good! Just some questions.

File: analysis_options.yaml [hunk 6]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - library_names
     - library_prefixes
     - list_remove_unrelated_type
+    - literal_only_boolean_expressions
+    - missing_whitespace_between_adjacent_strings
     - no_adjacent_strings_in_list
     - no_duplicate_case_values
-    - non_constant_identifier_names
+    - no_runtimeType_toString
+    - omit_local_variable_types
```
Comment (erickzanardo) [inline analysis_options.yaml:66]: Is this rule smart enough to detect cases where we need an empty variable? Because on that case we need to add the type otherwise the variable will be dynamic.

File: analysis_options.yaml (hunk 4 above)
Comment (erickzanardo) [reply analysis_options.yaml:76]: This moved to line 117

File: doc/examples/collision_detection/lib/main.dart [hunk 7]
```diff
@@ -194,7 +192,7 @@ class MyGame extends BaseGame with HasCollidables, HasDraggableComponents {
     MyCollidable lastToAdd = snowman;
     add(screen);
     add(snowman);
-    int totalAdded = 1;
+    var totalAdded = 1;
```
Comment (erickzanardo) [inline doc/examples/collision_detection/lib/main.dart:195]: This may be only my opinion, but I rather prefer the old style. Var just seems wrong to me hahaha, but this may be only me.

**Review by erickzanardo — APPROVED** [review]
Excellent, left a couple of comments, but LGTM already

File: analysis_options.yaml (hunk 6 above)
Comment (renancaraujo) [reply analysis_options.yaml:66]: IMO declaring the type is always better than `var` but that may be trauma from JS.

File: analysis_options.yaml [hunk 8]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - prefer_initializing_formals
+    - prefer_inlined_adds
+    - prefer_interpolation_to_compose_strings
     - prefer_is_empty
     - prefer_is_not_empty
+    - prefer_is_not_operator
+    - prefer_iterable_whereType
+    - prefer_mixin
+    - prefer_null_aware_operators
+    - prefer_single_quotes
```
Comment (renancaraujo) [inline analysis_options.yaml:99]: You should burn.

File: analysis_options.yaml [hunk 9]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - hash_and_equals
     - implementation_imports
+    - invariant_booleans
     - iterable_contains_unrelated_type
+    - join_return_with_assignment
     - library_names
     - library_prefixes
     - list_remove_unrelated_type
+    - literal_only_boolean_expressions
+    - missing_whitespace_between_adjacent_strings
```
Comment (renancaraujo) [inline analysis_options.yaml:62]: A little bit intrusive this one.

File: analysis_options.yaml [hunk 10]
```diff
@@ -15,69 +15,130 @@ linter:
...
     - empty_catches
     - empty_constructor_bodies
     - empty_statements
+    - exhaustive_cases
+    - file_names
     - hash_and_equals
     - implementation_imports
+    - invariant_booleans
     - iterable_contains_unrelated_type
+    - join_return_with_assignment
```
Comment (renancaraujo) [inline analysis_options.yaml:57]: I really hope this one has autofix

File: doc/examples/go_desktop/lib/game.dart [hunk 11]
```diff
@@ -6,10 +6,10 @@ import 'package:flutter/material.dart';
...
   static final Paint paint = Paint()..color = const Color(0xFFFFFFFF);
 
-  var movingLeft = false;
-  var movingRight = false;
-  var movingUp = false;
-  var movingDown = false;
+  bool movingLeft = false;
+  bool movingRight = false;
+  bool movingUp = false;
+  bool movingDown = false;
```
Comment (renancaraujo) [inline doc/examples/go_desktop/lib/game.dart:13]: This is the opposite of `omit_local_variable_types` (but in fields, not local vars). What rule required this change?

File: lib/src/components/joystick/joystick_action.dart [hunk 12]
```diff
@@ -16,7 +16,7 @@ import 'joystick_component.dart';
 import 'joystick_events.dart';
 import 'joystick_utils.dart';
 
-enum JoystickActionAlign { TOP_LEFT, BOTTOM_LEFT, TOP_RIGHT, BOTTOM_RIGHT }
+enum JoystickActionAlign { topLeft, bottomLeft, topRight, bottomRight }
```
Comment (renancaraujo) [inline lib/src/components/joystick/joystick_action.dart:19]: Breaking change

**Review by renancaraujo — APPROVED** [review]

File: analysis_options.yaml (hunk 2 above)
Comment (wolfenrain) [reply analysis_options.yaml:47]: These are the `--dart-define` environment variables, there was a whole threat about it on Flutter Study Group Slack. It breaks on null-safety code I believe.

File: analysis_options.yaml (hunk 1 above)
Comment (luanpotter) [reply analysis_options.yaml:21]: this page has the explanation: https://dart-lang.github.io/linter/lints/avoid_equals_and_hash_code_on_mutable_classes.html

basically it breaks let's say if you mutate the class while mid-sorting a list

nothing broke because of this rule, I just added to keep an eye in the future.

I am ok with removing if we think it doesn't make sense

File: analysis_options.yaml (hunk 3 above)
Comment (luanpotter) [reply analysis_options.yaml:40]: I changed it to `constant_identifier_names` (camel case names)

File: analysis_options.yaml (hunk 4 above)
Comment (luanpotter) [reply analysis_options.yaml:76]: yep just sorted alphabetically for making it easier to validate against the full list

File: doc/examples/sprites/lib/main_base64.dart (hunk 5 above)
Comment (luanpotter) [reply doc/examples/sprites/lib/main_base64.dart:27]: agree. but that can be a separate pr :)

File: analysis_options.yaml (hunk 6 above)
Comment (luanpotter) [reply analysis_options.yaml:66]: haha var in dart infers, unlike JS. same thing we already do for final. I think if it's not initialized then you are forced to use the type, but I will test rn

File: analysis_options.yaml (hunk 10 above)
Comment (luanpotter) [reply analysis_options.yaml:57]: it has (at least on vscode)

File: doc/examples/collision_detection/lib/main.dart (hunk 7 above)
Comment (luanpotter) [reply doc/examples/collision_detection/lib/main.dart:195]: var is the same as final that we already use. the rule I propose is: local variables are never explicitly typed. fields/methods are always explicitly typed.

in this case int is quite short but most of the time the type declaration would be pretty big and uncessary.

File: analysis_options.yaml (hunk 8 above)
Comment (luanpotter) [reply analysis_options.yaml:99]: :flame:

File: analysis_options.yaml (hunk 9 above)
Comment (luanpotter) [reply analysis_options.yaml:62]: I thought so too but could be helpful so catch errors. So far it has not triggered. if it does in a wrong scenario we can remove.

File: doc/examples/go_desktop/lib/game.dart (hunk 11 above)
Comment (luanpotter) [reply doc/examples/go_desktop/lib/game.dart:13]: I commented elsewhere, but basically:
local variables are never explicitly typed. fields/methods are always explicitly typed.

**Comment by luanpotter** [comment]
@wolfenrain yes mostly everything are the bullet points I listed

File: analysis_options.yaml (hunk 6 above)
Comment (luanpotter) [reply analysis_options.yaml:66]: yes if you don't initialize you are forced to type:

```
Prefer typing uninitialized variables and fields.dart(prefer_typing_uninitialized_variables)
```

so:
```
  var a = 1.0; // good
  double b; // good
  var c; // bad (would be dynamic)
  double d = 1.0; // bad (redundant)
```

Comment (erickzanardo) [reply analysis_options.yaml:66]: Sounds good then! I don't like car as well, but it seems smart enough.

Comment (luanpotter) [reply analysis_options.yaml:66]: We all know you only like motorbikes! :p 

Jokes aside, I don't like mutable variables at all. But they are a necessary evil. And the keyword var can bring terrible JS memories but kt/dark pretty much erased that dark past for me.

File: doc/examples/collision_detection/lib/main.dart (hunk 7 above)
Comment (erickzanardo) [reply doc/examples/collision_detection/lib/main.dart:195]: I know, but I think I am too hurt from JS so `var` always looks weird to me. But yeah, I like that rule

File: analysis_options.yaml (hunk 6 above)
Comment (erickzanardo) [reply analysis_options.yaml:66]: hahahha, I don't remember what I wanted to write there, I do like cars, but I do prefer bikes 🏍️

File: doc/examples/collision_detection/lib/main.dart (hunk 7 above)
Comment (spydon) [reply doc/examples/collision_detection/lib/main.dart:195]: I had exactly the same feeling as you Erick! But I think this will be fine, and it follows the dart styleguide.

File: analysis_options.yaml (hunk 1 above)
Comment (spydon) [reply analysis_options.yaml:21]: We can do the other way around, remove the rule it if it causes trouble.

**Review by spydon — APPROVED** [review]

File: analysis_options.yaml (hunk 1 above)
Comment (luanpotter) [reply analysis_options.yaml:21]: yeah I think that is a good rule of thumb!
